# Pandas
# file Read Write operation

In [1]:
import numpy as np
import pandas as pd


In [2]:
import pandas as pd 
draft1=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex1.csv')  #supply the file path  
draft1.head()

a   b     c     d message
0  1   2   3.0   4.0   hello
1  5   6   7.0   8.0   world
2  9  10  11.0  12.0     foo
3  4   5   NaN   NaN     NaN

In [3]:
type(draft1)

pandas.core.frame.DataFrame

In [4]:
draft1.head(2) # if we want two raws  from the top

a  b    c    d message
0  1  2  3.0  4.0   hello
1  5  6  7.0  8.0   world

In [5]:
draft1.tail(2) # if we want two raws  from the bottom

a   b     c     d message
2  9  10  11.0  12.0     foo
3  4   5   NaN   NaN     NaN

In [6]:
draft2=pd.read_table('/home/puru/Downloads/code_pandas1/examples/test.tsv')  #Read a tsv into a DataFrame
draft2.head()

test   test test.1 test.2 test.3
sudh    sudh   sudh   sudh   sudh   sudh
kumar  kumar  kumar  kumar  kumar  kumar

In [7]:
draft2=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/test.tsv',header=None)  #changed the parameter
draft2.head()

0
0             test\t test\ttest\ttest\ttest
1        sudh\tsudh\tsudh\tsudh\tsudh\tsudh
2  kumar\tkumar\tkumar\tkumar\tkumar\tkumar

In [8]:
# how to give own header name 
draft2=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex1.csv',names=["av","bv","er",'we'])
draft2.head()

av   bv   er       we
a   b    c    d  message
1   2    3    4    hello
5   6    7    8    world
9  10   11   12      foo
4   5  NaN  NaN      NaN

In [9]:
draft2=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/csv_mindex.csv',index_col=['key1','key2'])
draft2


value1  value2
key1 key2                
one  a          1       2
     b          3       4
     c          5       6
     d          7       8
two  a          9      10
     b         11      12
     c         13      14
     d         15      16

In [10]:
draft2=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/csv_mindex.csv')


In [11]:
draft2

key1 key2  value1  value2
0  one    a       1       2
1  one    b       3       4
2  one    c       5       6
3  one    d       7       8
4  two    a       9      10
5  two    b      11      12
6  two    c      13      14
7  two    d      15      16

In [12]:
#junk data in this data set
pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex4.csv')

# hey!
a                                                  b        c   d    message
# just wanted to make things more difficult for... NaN      NaN NaN      NaN
# who reads CSV files with computers                anyway? NaN NaN      NaN
1                                                  2        3   4      hello
5                                                  6        7   8      world
9                                                  10       11  12       foo

In [13]:
#how to remove junk data
pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex4.csv',skiprows=[0,2,3])

a   b   c   d message
0  1   2   3   4   hello
1  5   6   7   8   world
2  9  10  11  12     foo

In [14]:
  result=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv') #  datamissing value by deault it would be nan

In [15]:
result

Unnamed: 0  0
0  2000-01-01  0
1  2000-01-02  1
2  2000-01-03  2
3  2000-01-04  3
4  2000-01-05  4
5  2000-01-06  5
6  2000-01-07  6

In [16]:
result=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv') #  datamissing value by deault it would be nan
result
pd.isnull(result)  #identify where is null value it give us true

Unnamed: 0      0
0       False  False
1       False  False
2       False  False
3       False  False
4       False  False
5       False  False
6       False  False

In [17]:
pip install lxml

  Using cached https://files.pythonhosted.org/packages/79/37/d420b7fdc9a550bd29b8cfeacff3b38502d9600b09d7dfae9a69e623b891/lxml-4.5.2-cp36-cp36m-manylinux1_x86_64.whl
Note: you may need to restart the kernel to use updated packages.


In [18]:
#how to load data from url
url=  "https://www.basketball-reference.com/leagues/NBA_2016.html"
BB_data=pd.read_html(url) # read data from the specified url
BB_data[0] #called the datset in form of table

Eastern Conference   W   L   W/L%    GB   PS/G   PA/G   SRS
0   Cleveland Cavaliers*  57  25  0.695     —  104.3   98.3  5.45
1       Toronto Raptors*  56  26  0.683   1.0  102.7   98.2  4.08
2            Miami Heat*  48  34  0.585   9.0  100.0   98.4  1.50
3         Atlanta Hawks*  48  34  0.585   9.0  102.8   99.2  3.49
4        Boston Celtics*  48  34  0.585   9.0  105.7  102.5  2.84
5     Charlotte Hornets*  48  34  0.585   9.0  103.4  100.7  2.36
6        Indiana Pacers*  45  37  0.549  12.0  102.2  100.5  1.62
7       Detroit Pistons*  44  38  0.537  13.0  102.0  101.4  0.43
8          Chicago Bulls  42  40  0.512  15.0  101.6  103.1 -1.46
9     Washington Wizards  41  41  0.500  16.0  104.1  104.6 -0.50
10         Orlando Magic  35  47  0.427  22.0  102.1  103.7 -1.68
11       Milwaukee Bucks  33  49  0.402  24.0   99.0  103.2 -3.98
12       New York Knicks  32  50  0.390  25.0   98.4  101.1 -2.74
13         Brooklyn Nets  21  61  0.256  36.0   98.6  106.0 -7.12
14    Philadelphia 76ers  10  72  0.122  47.0   97.4  107.6 -9.92

In [19]:
#another datset
titanic_train=pd.read_csv("https://gist.githubusercontent.com/michhar/2dfd2de0d4f8727f873422c5d959fff5/raw/ff414a1bcfcba32481e4d4e8db578e55872a2ca1/titanic.csv",sep='\t')
titanic_train.head(5)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [20]:
titanic_train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [21]:
titanic_train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [22]:
titanic_train['Pclass']

0      3
1      1
2      3
3      1
4      3
      ..
151    1
152    3
153    3
154    3
155    1
Name: Pclass, Length: 156, dtype: int64

In [23]:
titanic_train['Pclass','Cabin']  # you are not able to do that 

KeyError: ('Pclass', 'Cabin')

In [ ]:
type(titanic_train['Pclass']) # if iselect one raw and one coumum then datatype will return series not dataframe

In [ ]:
# if we have to select multiple column in that case your suppose to provide all of those column in form of a list

In [ ]:
titanic_train[["Pclass","Survived"]]

In [ ]:
titanic_train[["Pclass","Survived","PassengerId"]]

In [ ]:
type(titanic_train[["Pclass","Survived","PassengerId"]])

In [ ]:
titanic_train[["Pclass","Survived","PassengerId"]].head(20)

In [ ]:
pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex6.csv',nrows=5)

In [ ]:
chunk=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex6.csv',chunksize=1000)  
type(chunk) of the data size means


Writing Data to Text Format

In [24]:
data=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv')
data

Unnamed: 0  0
0  2000-01-01  0
1  2000-01-02  1
2  2000-01-03  2
3  2000-01-04  3
4  2000-01-05  4
5  2000-01-06  5
6  2000-01-07  6

In [25]:
BB_data[0] # this data is already stor in my buffer memoray  whiuch is already inside my notebook itself

Eastern Conference   W   L   W/L%    GB   PS/G   PA/G   SRS
0   Cleveland Cavaliers*  57  25  0.695     —  104.3   98.3  5.45
1       Toronto Raptors*  56  26  0.683   1.0  102.7   98.2  4.08
2            Miami Heat*  48  34  0.585   9.0  100.0   98.4  1.50
3         Atlanta Hawks*  48  34  0.585   9.0  102.8   99.2  3.49
4        Boston Celtics*  48  34  0.585   9.0  105.7  102.5  2.84
5     Charlotte Hornets*  48  34  0.585   9.0  103.4  100.7  2.36
6        Indiana Pacers*  45  37  0.549  12.0  102.2  100.5  1.62
7       Detroit Pistons*  44  38  0.537  13.0  102.0  101.4  0.43
8          Chicago Bulls  42  40  0.512  15.0  101.6  103.1 -1.46
9     Washington Wizards  41  41  0.500  16.0  104.1  104.6 -0.50
10         Orlando Magic  35  47  0.427  22.0  102.1  103.7 -1.68
11       Milwaukee Bucks  33  49  0.402  24.0   99.0  103.2 -3.98
12       New York Knicks  32  50  0.390  25.0   98.4  101.1 -2.74
13         Brooklyn Nets  21  61  0.256  36.0   98.6  106.0 -7.12
14    Philadelphia 76ers  10  72  0.122  47.0   97.4  107.6 -9.92

In [26]:
BB_data[0].to_csv('NBA.csv')

In [27]:
BB_data[0].columns

Index(['Eastern Conference', 'W', 'L', 'W/L%', 'GB', 'PS/G', 'PA/G', 'SRS'], dtype='object')

In [28]:
import sys #input ,output stream

In [29]:
titanic_train.to_csv('Titanic.csv') # saved dataset in local directory 

In [30]:
titanic_train[["Name","Pclass"]].head()

Name  Pclass
0                            Braund, Mr. Owen Harris       3
1  Cumings, Mrs. John Bradley (Florence Briggs Th...       1
2                             Heikkinen, Miss. Laina       3
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1
4                           Allen, Mr. William Henry       3

In [31]:
titanic_train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
Age            float64
SibSp            int64
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
dtype: object

In [32]:
pd.options.display.max_rows = 10
result=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv')
result

Unnamed: 0  0
0  2000-01-01  0
1  2000-01-02  1
2  2000-01-03  2
3  2000-01-04  3
4  2000-01-05  4
5  2000-01-06  5
6  2000-01-07  6

In [33]:
dates=pd.date_range('1/1/2000',periods=7) #function
dates

DatetimeIndex(['2000-01-01', '2000-01-02', '2000-01-03', '2000-01-04',
               '2000-01-05', '2000-01-06', '2000-01-07'],
              dtype='datetime64[ns]', freq='D')

In [34]:
dates=pd.date_range('1/1/2000',periods=7)
ts=pd.Series(np.arange(7),index=dates)
ts.to_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv')
!cat /home/puru/Downloads/code_pandas1/examples/ex5.csv

,0
2000-01-01,0
2000-01-02,1
2000-01-03,2
2000-01-04,3
2000-01-05,4
2000-01-06,5
2000-01-07,6


In [35]:
ts=pd.Series(np.arange(7))

In [36]:
dates=pd.date_range('1/1/2000',periods=7)
ts=pd.Series(np.arange(7),index=dates)
ts.to_csv('/home/puru/Downloads/code_pandas1/examples/ex5.csv')
!cat /home/puru/Downloads/code_pandas1/examples/ex5.csv

,0
2000-01-01,0
2000-01-02,1
2000-01-03,2
2000-01-04,3
2000-01-05,4
2000-01-06,5
2000-01-07,6


In [37]:
range(7.0)

TypeError: 'float' object cannot be interpreted as an integer

In [38]:
np.arange(.1,7.0,.2)

array([0.1, 0.3, 0.5, 0.7, 0.9, 1.1, 1.3, 1.5, 1.7, 1.9, 2.1, 2.3, 2.5,
       2.7, 2.9, 3.1, 3.3, 3.5, 3.7, 3.9, 4.1, 4.3, 4.5, 4.7, 4.9, 5.1,
       5.3, 5.5, 5.7, 5.9, 6.1, 6.3, 6.5, 6.7, 6.9])

In [39]:
np.arange(7)

array([0, 1, 2, 3, 4, 5, 6])

In [40]:
type(np.arange(7))

numpy.ndarray

In [41]:
ts=pd.Series(np.arange(1,20))
ts

0      1
1      2
2      3
3      4
4      5
      ..
14    15
15    16
16    17
17    18
18    19
Length: 19, dtype: int64

In [42]:
ts[0]

1

In [43]:
ts[-1]  # here is not negative index u will not able to find in Series

KeyError: -1

In [44]:
import csv
f=open('/home/puru/Downloads/code_pandas1/examples/ex7.csv')

reader=csv.reader(f)
reader

In [45]:
for line in reader:
    print(line)

['a', 'b', 'c']
['1', '2', '3']
['1', '2', '3']


In [46]:
with open("/home/puru/Downloads/code_pandas1/examples/ex7.csv") as f:
    lines=list(csv.reader(f))

In [47]:
lines

[['a', 'b', 'c'], ['1', '2', '3'], ['1', '2', '3']]

In [48]:
#zip function how does it work

In [49]:
header, values = lines[0], lines[1:] #declare two variable

In [50]:
header

['a', 'b', 'c']

In [51]:
values

[['1', '2', '3'], ['1', '2', '3']]

In [52]:
z= zip(header,zip(*values))

In [53]:
for i in z:
    print(i)

('a', ('1', '1'))
('b', ('2', '2'))
('c', ('3', '3'))


In [54]:
data_dict = {h: v for h, v in zip(header, zip(*values))} # zip always return tuples

In [55]:
for i in zip([1,2,3],[5,6,7]): # paired with indexes 
    print(i)

(1, 5)
(2, 6)
(3, 7)


In [56]:
d={}
for i in zip(header, zip(*values)):
    d[i[0]] =i[1]
d

{'a': ('1', '1'), 'b': ('2', '2'), 'c': ('3', '3')}

JSON Data

In [57]:
obj = """
{"name": "Wes",
 "places_lived": ["United States", "Spain", "Germany"],
 "pet": null,
 "siblings": [{"name": "Scott", "age": 30, "pets": ["Zeus", "Zuko"]},
              {"name": "Katie", "age": 38,"pets": ["Sixes", "Stache", "Cisco"]},
               {"name": "Puru", "age": 23,"pets": ["werrrr", "sgdhhdtdtd", "sssss"]}]
}
"""

In [58]:
import json
result =json.loads(obj)
result

{'name': 'Wes',
 'places_lived': ['United States', 'Spain', 'Germany'],
 'pet': None,
 'siblings': [{'name': 'Scott', 'age': 30, 'pets': ['Zeus', 'Zuko']},
  {'name': 'Katie', 'age': 38, 'pets': ['Sixes', 'Stache', 'Cisco']},
  {'name': 'Puru', 'age': 23, 'pets': ['werrrr', 'sgdhhdtdtd', 'sssss']}]}

In [59]:
type(obj)

str

In [60]:
type(result)

dict

In [61]:
asjson=json.dumps(result)

In [62]:
(asjson)

'{"name": "Wes", "places_lived": ["United States", "Spain", "Germany"], "pet": null, "siblings": [{"name": "Scott", "age": 30, "pets": ["Zeus", "Zuko"]}, {"name": "Katie", "age": 38, "pets": ["Sixes", "Stache", "Cisco"]}, {"name": "Puru", "age": 23, "pets": ["werrrr", "sgdhhdtdtd", "sssss"]}]}'

In [63]:
# how to convert json data into dataframe

In [64]:
siblings =pd.DataFrame(result["siblings"],columns=['name','age','pets'])
siblings

name  age                         pets
0  Scott   30                 [Zeus, Zuko]
1  Katie   38       [Sixes, Stache, Cisco]
2   Puru   23  [werrrr, sgdhhdtdtd, sssss]

In [65]:
result['siblings'] # all the key which is availbale in the dictionary

[{'name': 'Scott', 'age': 30, 'pets': ['Zeus', 'Zuko']},
 {'name': 'Katie', 'age': 38, 'pets': ['Sixes', 'Stache', 'Cisco']},
 {'name': 'Puru', 'age': 23, 'pets': ['werrrr', 'sgdhhdtdtd', 'sssss']}]

In [66]:
siblings = pd.DataFrame(result['siblings'], columns=['name', 'age','pets'])
siblings

name  age                         pets
0  Scott   30                 [Zeus, Zuko]
1  Katie   38       [Sixes, Stache, Cisco]
2   Puru   23  [werrrr, sgdhhdtdtd, sssss]

In [67]:
data=pd.read_json("/home/puru/Downloads/code_pandas1/examples/example.json") 
data

a  b  c
0  1  2  3
1  4  5  6
2  7  8  9

In [68]:
print(data.to_json())   
print(data.to_json(orient='records'))

{"a":{"0":1,"1":4,"2":7},"b":{"0":2,"1":5,"2":8},"c":{"0":3,"1":6,"2":9}}
[{"a":1,"b":2,"c":3},{"a":4,"b":5,"c":6},{"a":7,"b":8,"c":9}]


In [69]:
type(data.to_json())

str

XNL and HTML: Web Scraping

In [70]:
tables=pd.read_html('/home/puru/Downloads/code_pandas1/examples/fdic_failed_bank_list.html')
len(tables)
failures= tables[0]
failures.head()

Bank Name             City  ST   CERT  \
0                   Allied Bank         Mulberry  AR     91   
1  The Woodbury Banking Company         Woodbury  GA  11297   
2        First CornerStone Bank  King of Prussia  PA  35312   
3            Trust Company Bank          Memphis  TN   9956   
4    North Milwaukee State Bank        Milwaukee  WI  20364   

                 Acquiring Institution        Closing Date       Updated Date  
0                         Today's Bank  September 23, 2016  November 17, 2016  
1                          United Bank     August 19, 2016  November 17, 2016  
2  First-Citizens Bank & Trust Company         May 6, 2016  September 6, 2016  
3           The Bank of Fayette County      April 29, 2016  September 6, 2016  
4  First-Citizens Bank & Trust Company      March 11, 2016      June 16, 2016

In [71]:
failures.dtypes

Bank Name                object
City                     object
ST                       object
CERT                      int64
Acquiring Institution    object
Closing Date             object
Updated Date             object
dtype: object

In [72]:
failures["Closing Date"]

0      September 23, 2016
1         August 19, 2016
2             May 6, 2016
3          April 29, 2016
4          March 11, 2016
              ...        
542         July 27, 2001
543           May 3, 2001
544      February 2, 2001
545     December 14, 2000
546      October 13, 2000
Name: Closing Date, Length: 547, dtype: object

In [73]:
close_timestamps =pd.to_datetime(failures['Closing Date'])
close_timestamps.dt.year.value_counts()
close_timestamps

0     2016-09-23
1     2016-08-19
2     2016-05-06
3     2016-04-29
4     2016-03-11
         ...    
542   2001-07-27
543   2001-05-03
544   2001-02-02
545   2000-12-14
546   2000-10-13
Name: Closing Date, Length: 547, dtype: datetime64[ns]

# Binary Data Formats

In [74]:
frame=pd.read_csv('/home/puru/Downloads/code_pandas1/examples/ex1.csv')


In [75]:
frame

a   b     c     d message
0  1   2   3.0   4.0   hello
1  5   6   7.0   8.0   world
2  9  10  11.0  12.0     foo
3  4   5   NaN   NaN     NaN

In [76]:
import requests
url = 'https://api.github.com/repos/pandas-dev/pandas/issues'
resp = requests.get(url)
resp

<Response [200]>

In [77]:
data=resp.json()
type(data)
data[2]

{'url': 'https://api.github.com/repos/pandas-dev/pandas/issues/35654',
 'repository_url': 'https://api.github.com/repos/pandas-dev/pandas',
 'labels_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/35654/labels{/name}',
 'comments_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/35654/comments',
 'events_url': 'https://api.github.com/repos/pandas-dev/pandas/issues/35654/events',
 'html_url': 'https://github.com/pandas-dev/pandas/pull/35654',
 'id': 676208659,
 'node_id': 'MDExOlB1bGxSZXF1ZXN0NDY1NTYwMDc1',
 'number': 35654,
 'title': 'BUG: GH-35558 merge_asof tolerance error',
 'user': {'login': 'ikedaosushi',
  'id': 4068610,
  'node_id': 'MDQ6VXNlcjQwNjg2MTA=',
  'avatar_url': 'https://avatars1.githubusercontent.com/u/4068610?v=4',
  'gravatar_id': '',
  'url': 'https://api.github.com/users/ikedaosushi',
  'html_url': 'https://github.com/ikedaosushi',
  'followers_url': 'https://api.github.com/users/ikedaosushi/followers',
  'following_url': 'https://api.githu

In [78]:
issues = pd.DataFrame(data, columns=['number', 'title',
                                     'labels', 'state'])
issues

number                                              title  \
0    35656     ENH: Uniformised the boolean results in Pandas   
1    35655                                        Moto server   
2    35654           BUG: GH-35558 merge_asof tolerance error   
3    35653  QST: How to select the week data from a given ...   
4    35650                                               BUG:   
..     ...                                                ...   
25   35620  CI: Linux py36_locale failures with pytest Dep...   
26   35616  QST:TypeError: sum() got an unexpected keyword...   
27   35614             ENH: Make Series.explode work for sets   
28   35613   REF: use consistent pattern in tslibs.vectorized   
29   35612  BUG: DataFrameGroupBy.__getitem__ fails to pro...   

                                               labels state  
0   [{'id': 76812, 'node_id': 'MDU6TGFiZWw3NjgxMg=...  open  
1                                                  []  open  
2                                                  []  open  
3   [{'id': 1954720290, 'node_id': 'MDU6TGFiZWwxOT...  open  
4   [{'id': 76811, 'node_id': 'MDU6TGFiZWw3NjgxMQ=...  open  
..                                                ...   ...  
25  [{'id': 48070600, 'node_id': 'MDU6TGFiZWw0ODA3...  open  
26  [{'id': 1954720290, 'node_id': 'MDU6TGFiZWwxOT...  open  
27  [{'id': 76812, 'node_id': 'MDU6TGFiZWw3NjgxMg=...  open  
28  [{'id': 127681, 'node_id': 'MDU6TGFiZWwxMjc2OD...  open  
29  [{'id': 76811, 'node_id': 'MDU6TGFiZWw3NjgxMQ=...  open  

[30 rows x 4 columns]

In [79]:
issues = pd.DataFrame(data)
issues.head(3)


url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   

                                   repository_url  \
0  https://api.github.com/repos/pandas-dev/pandas   
1  https://api.github.com/repos/pandas-dev/pandas   
2  https://api.github.com/repos/pandas-dev/pandas   

                                          labels_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   

                                        comments_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   

                                          events_url  \
0  https://api.github.com/repos/pandas-dev/pandas...   
1  https://api.github.com/repos/pandas-dev/pandas...   
2  https://api.github.com/repos/pandas-dev/pandas...   

                                            html_url         id  \
0  https://github.com/pandas-dev/pandas/issues/35656  676242685   
1    https://github.com/pandas-dev/pandas/pull/35655  676232008   
2    https://github.com/pandas-dev/pandas/pull/35654  676208659   

                            node_id  number  \
0          MDU6SXNzdWU2NzYyNDI2ODU=   35656   
1  MDExOlB1bGxSZXF1ZXN0NDY1NTc5NzEx   35655   
2  MDExOlB1bGxSZXF1ZXN0NDY1NTYwMDc1   35654   

                                            title  ... milestone comments  \
0  ENH: Uniformised the boolean results in Pandas  ...      None        0   
1                                     Moto server  ...      None        0   
2        BUG: GH-35558 merge_asof tolerance error  ...      None        0   

             created_at            updated_at closed_at author_association  \
0  2020-08-10T15:56:08Z  2020-08-10T15:56:08Z      None               NONE   
1  2020-08-10T15:40:01Z  2020-08-10T16:04:15Z      None        CONTRIBUTOR   
2  2020-08-10T15:07:53Z  2020-08-10T15:08:29Z      None               NONE   

  active_lock_reason                                               body  \
0               None  Hello everybody,\r\n\r\nI'm not completely sur...   
1               None  Changes moto for s3 tests from monkeypatched/m...   
2               None  - [x] closes #35558\r\n- [x] tests added / pas...   

  performed_via_github_app                                       pull_request  
0                     None                                                NaN  
1                     None  {'url': 'https://api.github.com/repos/pandas-d...  
2                     None  {'url': 'https://api.github.com/repos/pandas-d...  

[3 rows x 26 columns]

In [80]:
issues['node_id']

0             MDU6SXNzdWU2NzYyNDI2ODU=
1     MDExOlB1bGxSZXF1ZXN0NDY1NTc5NzEx
2     MDExOlB1bGxSZXF1ZXN0NDY1NTYwMDc1
3             MDU6SXNzdWU2NzYyMDI1MzM=
4             MDU6SXNzdWU2NzU5ODc4MzQ=
                    ...               
25            MDU6SXNzdWU2NzU0OTEzNzI=
26            MDU6SXNzdWU2NzUzMjc2NDc=
27            MDU6SXNzdWU2NzUzMTMyNDc=
28    MDExOlB1bGxSZXF1ZXN0NDY0ODQyMjc5
29            MDU6SXNzdWU2NzUyMjQxMDQ=
Name: node_id, Length: 30, dtype: object

# Interacting with Databases

In [81]:
import sqlite3
#connecting with the database.
db = sqlite3.connect("my_database4.db")
# Drop table if it already exist using execute() method.
db.execute("drop table if exists test")
query = """
CREATE TABLE test
(a VARCHAR(20), b VARCHAR(20),
 c REAL,        d INTEGER
);"""
con = sqlite3.connect('mydata2.sqlite')
con.execute(query)
con.commit()

OperationalError: table test already exists

In [82]:
data = [('Atlanta', 'Georgia', 1.25, 6),
        ('Tallahassee', 'Florida', 2.6, 3),
        ('Sacramento', 'California', 1.7, 5)]
stmt = "INSERT INTO test VALUES(?, ?, ?, ?)"
con.executemany(stmt, data)
con.commit()

In [83]:
cursor = con.execute('select * from test')
rows = cursor.fetchall()
rows

[('Atlanta', 'Georgia', 1.25, 6),
 ('Tallahassee', 'Florida', 2.6, 3),
 ('Sacramento', 'California', 1.7, 5),
 ('Atlanta', 'Georgia', 1.25, 6),
 ('Tallahassee', 'Florida', 2.6, 3),
 ('Sacramento', 'California', 1.7, 5)]

In [84]:
cursor.description
pd.DataFrame(rows, columns=[x[0] for x in cursor.description])

a           b     c  d
0      Atlanta     Georgia  1.25  6
1  Tallahassee     Florida  2.60  3
2   Sacramento  California  1.70  5
3      Atlanta     Georgia  1.25  6
4  Tallahassee     Florida  2.60  3
5   Sacramento  California  1.70  5

# Pandas data manipulation 

In [85]:
import pandas as pd

In [86]:
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
%matplotlib inline

In [87]:
titanic_train = pd.read_csv("https://gist.githubusercontent.com/michhar/2dfd2de0d4f8727f873422c5d959fff5/raw/ff414a1bcfcba32481e4d4e8db578e55872a2ca1/titanic.csv",
                           sep='\t')   
titanic_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
151      0            113776  66.6000    C2        S  
152      0        A.5. 11206   8.0500   NaN        S  
153      2          A/5. 851  14.5000   NaN        S  
154      0         Fa 265302   7.3125   NaN        S  
155      1          PC 17597  61.3792   NaN        C  

[156 rows x 12 columns]

In [88]:
titanic_train.dtypes

PassengerId      int64
Survived         int64
Pclass           int64
Name            object
Sex             object
                ...   
Parch            int64
Ticket          object
Fare           float64
Cabin           object
Embarked        object
Length: 12, dtype: object

In [89]:
titanic_train.dtypes =="objects"  # datatype is equal to object

PassengerId    False
Survived       False
Pclass         False
Name           False
Sex            False
               ...  
Parch          False
Ticket         False
Fare           False
Cabin          False
Embarked       False
Length: 12, dtype: bool

In [90]:
type(titanic_train.dtypes =="objects")

pandas.core.series.Series

In [91]:
(titanic_train.dtypes =="int64")# i have to filter out all column which was having a object dtype

PassengerId     True
Survived        True
Pclass          True
Name           False
Sex            False
               ...  
Parch           True
Ticket         False
Fare           False
Cabin          False
Embarked       False
Length: 12, dtype: bool

In [92]:
titanic_train.dtypes[titanic_train.dtypes =="int64"] #based on the datatype i am filter it out 

PassengerId    int64
Survived       int64
Pclass         int64
SibSp          int64
Parch          int64
dtype: object

In [93]:
titanic_train.dtypes == "object"   
a = titanic_train.dtypes[titanic_train.dtypes == "object"].index
a


Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [94]:
titanic_train.dtypes[titanic_train.dtypes == "object"].index # only index no data

Index(['Name', 'Sex', 'Ticket', 'Cabin', 'Embarked'], dtype='object')

In [95]:
titanic_train[['Embarked']] #how to extract Embarked class column

Embarked
0          S
1          C
2          S
3          S
4          S
..       ...
151        S
152        S
153        S
154        S
155        C

[156 rows x 1 columns]

here you will be able to find now that it is trying to give you a count means how many data sets are available in
a name column 156 in to a ticket coloumn 156. there are some data set which is available inside a keeping column 
which can have a value is equal to n a n means a null value so thats the reason so it is not equal 156 so this count opeartion always try to return the number of the data and it always excludes nan values.

In [96]:
titanic_train[a].describe() 


Name   Sex Ticket        Cabin Embarked
count                   156   156    156           31      155
unique                  156     2    145           28        3
top     Todoroff, Mr. Lalio  male  19950  C23 C25 C27        S
freq                      1   100      2            2      110

In [97]:
titanic_train.describe() # this is going to give you data which is having a numerical natural


PassengerId    Survived      Pclass         Age       SibSp  \
count   156.000000  156.000000  156.000000  126.000000  156.000000   
mean     78.500000    0.346154    2.423077   28.141508    0.615385   
std      45.177428    0.477275    0.795459   14.613880    1.056235   
min       1.000000    0.000000    1.000000    0.830000    0.000000   
25%      39.750000    0.000000    2.000000   19.000000    0.000000   
50%      78.500000    0.000000    3.000000   26.000000    0.000000   
75%     117.250000    1.000000    3.000000   35.000000    1.000000   
max     156.000000    1.000000    3.000000   71.000000    5.000000   

            Parch        Fare  
count  156.000000  156.000000  
mean     0.397436   28.109587  
std      0.870146   39.401047  
min      0.000000    6.750000  
25%      0.000000    8.003150  
50%      0.000000   14.454200  
75%      0.000000   30.371850  
max      5.000000  263.000000

In [98]:
titanic_train['Name']

0                                Braund, Mr. Owen Harris
1      Cumings, Mrs. John Bradley (Florence Briggs Th...
2                                 Heikkinen, Miss. Laina
3           Futrelle, Mrs. Jacques Heath (Lily May Peel)
4                               Allen, Mr. William Henry
                             ...                        
151                    Pears, Mrs. Thomas (Edith Wearne)
152                                     Meo, Mr. Alfonzo
153                      van Billiard, Mr. Austin Blyler
154                                Olsen, Mr. Ole Martin
155                          Williams, Mr. Charles Duane
Name: Name, Length: 156, dtype: object

In [99]:
sorted(titanic_train['Name'])

['Ahlin, Mrs. Johan (Johanna Persdotter Larsson)',
 'Allen, Mr. William Henry',
 'Andersson, Miss. Ellis Anna Maria',
 'Andersson, Miss. Erna Alexandra',
 'Andersson, Mr. Anders Johan',
 'Andersson, Mr. August Edvard ("Wennerstrom")',
 'Andreasson, Mr. Paul Edvin',
 'Andrew, Mr. Edgardo Samuel',
 'Arnold-Franchi, Mrs. Josef (Josefine Franchi)',
 'Asplund, Mrs. Carl Oscar (Selma Augusta Emilia Johansson)',
 'Attalah, Miss. Malake',
 'Backstrom, Mrs. Karl Alfred (Maria Mathilda Gustafsson)',
 'Barton, Mr. David John',
 'Bateman, Rev. Robert James',
 'Baxter, Mr. Quigg Edmond',
 'Beesley, Mr. Lawrence',
 'Bing, Mr. Lee',
 'Bonnell, Miss. Elizabeth',
 'Boulos, Mrs. Joseph (Sultana)',
 'Braund, Mr. Owen Harris',
 'Burke, Mr. Jeremiah',
 'Byles, Rev. Thomas Roussel Davids',
 'Caldwell, Master. Alden Gates',
 'Cann, Mr. Ernest Charles',
 'Carrau, Mr. Francisco M',
 'Celotti, Mr. Francesco',
 'Chaffee, Mr. Herbert Fuller',
 'Christmann, Mr. Emil',
 'Chronopoulos, Mr. Apostolos',
 'Coelho, Mr. 

In [100]:
(sorted(titanic_train['Name']))[0:5]  # o th index till 5th indexes

['Ahlin, Mrs. Johan (Johanna Persdotter Larsson)',
 'Allen, Mr. William Henry',
 'Andersson, Miss. Ellis Anna Maria',
 'Andersson, Miss. Erna Alexandra',
 'Andersson, Mr. Anders Johan']

In [101]:
(sorted(titanic_train['Name']))[0:90:5]  # data starting from a zero to may be 90th position and i wanted to get maybe like every fifth data

['Ahlin, Mrs. Johan (Johanna Persdotter Larsson)',
 'Andersson, Mr. August Edvard ("Wennerstrom")',
 'Attalah, Miss. Malake',
 'Beesley, Mr. Lawrence',
 'Burke, Mr. Jeremiah',
 'Celotti, Mr. Francesco',
 'Connors, Mr. Patrick',
 'Devaney, Miss. Margaret Delia',
 'Emir, Mr. Farred Chehab',
 'Fortune, Mr. Charles Alexander',
 'Glynn, Miss. Mary Agatha',
 'Gustafsson, Mr. Anders Vilhelm',
 'Hewlett, Mrs. (Mary D Kingcome) ',
 'Ilett, Miss. Bertha',
 'Kantor, Mr. Sinai',
 'Madsen, Mr. Fridtjof Arne',
 'McGowan, Miss. Anna "Annie"',
 'Moen, Mr. Sigurd Hansen']

In [102]:
(sorted(titanic_train['Name']))[::-1]  # data into descending order

['van Billiard, Mr. Austin Blyler',
 'Zabour, Miss. Hileni',
 'Woolner, Mr. Hugh',
 'Williams, Mr. Charles Eugene',
 'Williams, Mr. Charles Duane',
 'White, Mr. Richard Frasar',
 'White, Mr. Percival Wayland',
 'Wheadon, Mr. Edward H',
 'West, Miss. Constance Mirium',
 'Weisz, Mrs. Leopold (Mathilde Francoise Pede)',
 'Webber, Miss. Susan',
 'Waelens, Mr. Achille',
 'Vestrom, Miss. Hulda Amanda Adolfina',
 'Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)',
 'Vander Planke, Miss. Augusta Maria',
 'Uruchurtu, Don. Manuel E',
 'Turpin, Mrs. William John Robert (Dorothy Ann Wonnacott)',
 'Turpin, Mr. William John Robert',
 'Todoroff, Mr. Lalio',
 'Stewart, Mr. Albert A',
 'Staneff, Mr. Ivan',
 'Spencer, Mrs. William Augustus (Marie Eugenie)',
 'Sobey, Mr. Samuel James Hayden',
 'Sloper, Mr. William Thompson',
 'Slocovski, Mr. Selman Francis',
 'Skoog, Master. Harald',
 'Sirayanian, Mr. Orsen',
 'Shorney, Mr. Charles Joseph',
 'Sheerlinck, Mr. Jan Baptist',
 'Saundercock, Mr. Willia

In [103]:
titanic_train.head()

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [104]:
titanic_train[['Name']][4:7] # starting from fourth index to 7 th index excluding 7yh index
# in this name section we can pass any name of column

Name
4  Allen, Mr. William Henry
5          Moran, Mr. James
6   McCarthy, Mr. Timothy J

In [105]:
titanic_train.columns

Index(['PassengerId', 'Survived', 'Pclass', 'Name', 'Sex', 'Age', 'SibSp',
       'Parch', 'Ticket', 'Fare', 'Cabin', 'Embarked'],
      dtype='object')

In [106]:
titanic_train[['Name','Sex','Age']]

Name     Sex   Age
0                              Braund, Mr. Owen Harris    male  22.0
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0
2                               Heikkinen, Miss. Laina  female  26.0
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0
4                             Allen, Mr. William Henry    male  35.0
..                                                 ...     ...   ...
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0
152                                   Meo, Mr. Alfonzo    male  55.5
153                    van Billiard, Mr. Austin Blyler    male  40.5
154                              Olsen, Mr. Ole Martin    male   NaN
155                        Williams, Mr. Charles Duane    male  51.0

[156 rows x 3 columns]

In [107]:
pd.Categorical(titanic_train["Pclass"])

[3, 1, 3, 1, 3, ..., 1, 3, 3, 3, 1]
Length: 156
Categories (3, int64): [1, 2, 3]

In [108]:
a=pd.Categorical(titanic_train["Pclass"])
b=a.rename_categories(["class1","class2","class3"])
b
titanic_train["newcol"]=b

In [109]:
titanic_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     Parch            Ticket     Fare Cabin Embarked  newcol  
0        0         A/5 21171   7.2500   NaN        S  class3  
1        0          PC 17599  71.2833   C85        C  class1  
2        0  STON/O2. 3101282   7.9250   NaN        S  class3  
3        0            113803  53.1000  C123        S  class1  
4        0            373450   8.0500   NaN        S  class3  
..     ...               ...      ...   ...      ...     ...  
151      0            113776  66.6000    C2        S  class1  
152      0        A.5. 11206   8.0500   NaN        S  class3  
153      2          A/5. 851  14.5000   NaN        S  class3  
154      0         Fa 265302   7.3125   NaN        S  class3  
155      1          PC 17597  61.3792   NaN        C  class1  

[156 rows x 13 columns]

In [110]:
titanic_train['Name'].unique()

array(['Braund, Mr. Owen Harris',
       'Cumings, Mrs. John Bradley (Florence Briggs Thayer)',
       'Heikkinen, Miss. Laina',
       'Futrelle, Mrs. Jacques Heath (Lily May Peel)',
       'Allen, Mr. William Henry', 'Moran, Mr. James',
       'McCarthy, Mr. Timothy J', 'Palsson, Master. Gosta Leonard',
       'Johnson, Mrs. Oscar W (Elisabeth Vilhelmina Berg)',
       'Nasser, Mrs. Nicholas (Adele Achem)',
       'Sandstrom, Miss. Marguerite Rut', 'Bonnell, Miss. Elizabeth',
       'Saundercock, Mr. William Henry', 'Andersson, Mr. Anders Johan',
       'Vestrom, Miss. Hulda Amanda Adolfina',
       'Hewlett, Mrs. (Mary D Kingcome) ', 'Rice, Master. Eugene',
       'Williams, Mr. Charles Eugene',
       'Vander Planke, Mrs. Julius (Emelia Maria Vandemoortele)',
       'Masselmani, Mrs. Fatima', 'Fynney, Mr. Joseph J',
       'Beesley, Mr. Lawrence', 'McGowan, Miss. Anna "Annie"',
       'Sloper, Mr. William Thompson', 'Palsson, Miss. Torborg Danira',
       'Asplund, Mrs. Carl Oscar 

In [111]:
titanic_train["Cabin"].unique()

array([nan, 'C85', 'C123', 'E46', 'G6', 'C103', 'D56', 'A6',
       'C23 C25 C27', 'B78', 'D33', 'B30', 'C52', 'B28', 'C83', 'F33',
       'F G73', 'E31', 'A5', 'D10 D12', 'D26', 'C110', 'B58 B60', 'E101',
       'F E69', 'D47', 'B86', 'F2', 'C2'], dtype=object)

In [112]:
titanic_train["Cabin"].head(20)

0      NaN
1      C85
2      NaN
3     C123
4      NaN
      ... 
15     NaN
16     NaN
17     NaN
18     NaN
19     NaN
Name: Cabin, Length: 20, dtype: object

In [113]:
titanic_train["Cabin"].dtypes

dtype('O')

In [114]:
char_cabin= titanic_train['Cabin'].astype(str)
for i in char_cabin:
    print(i)

nan
C85
nan
C123
nan
nan
E46
nan
nan
nan
G6
C103
nan
nan
nan
nan
nan
nan
nan
nan
nan
D56
nan
A6
nan
nan
nan
C23 C25 C27
nan
nan
nan
B78
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
D33
nan
B30
C52
nan
nan
nan
nan
nan
B28
C83
nan
nan
nan
F33
nan
nan
nan
nan
nan
nan
nan
nan
F G73
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
nan
C23 C25 C27
nan
nan
nan
E31
nan
nan
nan
A5
D10 D12
nan
nan
nan
nan
D26
nan
nan
nan
nan
nan
nan
nan
C110
nan
nan
nan
nan
nan
nan
nan
B58 B60
nan
nan
nan
nan
E101
D26
nan
nan
nan
F E69
nan
nan
nan
nan
nan
nan
nan
D47
C123
nan
B86
nan
nan
nan
nan
nan
nan
nan
nan
F2
nan
nan
C2
nan
nan
nan
nan


In [115]:
char_cabin=titanic_train["Cabin"].astype(str) #take first letter from dataset
[i[0] for i in char_cabin]

['n',
 'C',
 'n',
 'C',
 'n',
 'n',
 'E',
 'n',
 'n',
 'n',
 'G',
 'C',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'D',
 'n',
 'A',
 'n',
 'n',
 'n',
 'C',
 'n',
 'n',
 'n',
 'B',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'D',
 'n',
 'B',
 'C',
 'n',
 'n',
 'n',
 'n',
 'n',
 'B',
 'C',
 'n',
 'n',
 'n',
 'F',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'F',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'C',
 'n',
 'n',
 'n',
 'E',
 'n',
 'n',
 'n',
 'A',
 'D',
 'n',
 'n',
 'n',
 'n',
 'D',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'C',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'B',
 'n',
 'n',
 'n',
 'n',
 'E',
 'D',
 'n',
 'n',
 'n',
 'F',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'D',
 'C',
 'n',
 'B',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'n',
 'F',
 'n',
 'n',
 'C',
 'n',
 'n',
 'n',
 'n']

In [116]:
char_cabin=titanic_train["Cabin"].astype(str) # convert data to str
new_Cabin = np.array([cabin[0] for cabin in char_cabin]) # take first letter
new_cabin=pd.Categorical(new_Cabin)
new_Cabin 

array(['n', 'C', 'n', 'C', 'n', 'n', 'E', 'n', 'n', 'n', 'G', 'C', 'n',
       'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'D', 'n', 'A', 'n', 'n',
       'n', 'C', 'n', 'n', 'n', 'B', 'n', 'n', 'n', 'n', 'n', 'n', 'n',
       'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n',
       'D', 'n', 'B', 'C', 'n', 'n', 'n', 'n', 'n', 'B', 'C', 'n', 'n',
       'n', 'F', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'F', 'n', 'n',
       'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'n', 'C', 'n', 'n',
       'n', 'E', 'n', 'n', 'n', 'A', 'D', 'n', 'n', 'n', 'n', 'D', 'n',
       'n', 'n', 'n', 'n', 'n', 'n', 'C', 'n', 'n', 'n', 'n', 'n', 'n',
       'n', 'B', 'n', 'n', 'n', 'n', 'E', 'D', 'n', 'n', 'n', 'F', 'n',
       'n', 'n', 'n', 'n', 'n', 'n', 'D', 'C', 'n', 'B', 'n', 'n', 'n',
       'n', 'n', 'n', 'n', 'n', 'F', 'n', 'n', 'C', 'n', 'n', 'n', 'n'],
      dtype='<U1')

In [117]:
dummy_vector =pd.Series([1,None,3,None,7,8])
dummy_vector 

0    1.0
1    NaN
2    3.0
3    NaN
4    7.0
5    8.0
dtype: float64

In [118]:
dummy_vector.isnull()

0    False
1     True
2    False
3     True
4    False
5    False
dtype: bool

In [119]:
# selection operation


In [120]:
titanic_train['Age']

0      22.0
1      38.0
2      26.0
3      35.0
4      35.0
       ... 
151    22.0
152    55.5
153    40.5
154     NaN
155    51.0
Name: Age, Length: 156, dtype: float64

In [121]:
titanic_train['Age'].isnull()

0      False
1      False
2      False
3      False
4      False
       ...  
151    False
152    False
153    False
154     True
155    False
Name: Age, Length: 156, dtype: bool

In [122]:
missing=np.where(titanic_train["Age"].isnull() == True) # this is going to return me a indexes wherever it is going to true

In [123]:
missing


(array([  5,  17,  19,  26,  28,  29,  31,  32,  36,  42,  45,  46,  47,
         48,  55,  64,  65,  76,  77,  82,  87,  95, 101, 107, 109, 121,
        126, 128, 140, 154]),)

In [124]:
len(missing[0]) # the total number of missing data

30

In [125]:
titanic_train.loc[missing]

PassengerId  Survived  Pclass                           Name     Sex  \
5              6         0       3               Moran, Mr. James    male   
17            18         1       2   Williams, Mr. Charles Eugene    male   
19            20         1       3        Masselmani, Mrs. Fatima  female   
26            27         0       3        Emir, Mr. Farred Chehab    male   
28            29         1       3  O'Dwyer, Miss. Ellen "Nellie"  female   
..           ...       ...     ...                            ...     ...   
121          122         0       3     Moore, Mr. Leonard Charles    male   
126          127         0       3            McMahon, Mr. Martin    male   
128          129         1       3              Peter, Miss. Anna  female   
140          141         0       3  Boulos, Mrs. Joseph (Sultana)  female   
154          155         0       3          Olsen, Mr. Ole Martin    male   

     Age  SibSp  Parch     Ticket     Fare  Cabin Embarked  newcol  
5    NaN      0      0     330877   8.4583    NaN        Q  class3  
17   NaN      0      0     244373  13.0000    NaN        S  class2  
19   NaN      0      0       2649   7.2250    NaN        C  class3  
26   NaN      0      0       2631   7.2250    NaN        C  class3  
28   NaN      0      0     330959   7.8792    NaN        Q  class3  
..   ...    ...    ...        ...      ...    ...      ...     ...  
121  NaN      0      0  A4. 54510   8.0500    NaN        S  class3  
126  NaN      0      0     370372   7.7500    NaN        Q  class3  
128  NaN      1      1       2668  22.3583  F E69        C  class3  
140  NaN      0      2       2678  15.2458    NaN        C  class3  
154  NaN      0      0  Fa 265302   7.3125    NaN        S  class3  

[30 rows x 13 columns]

In [126]:
titanic_train

PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
151          152         1       1   
152          153         0       3   
153          154         0       3   
154          155         0       3   
155          156         0       1   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
151                  Pears, Mrs. Thomas (Edith Wearne)  female  22.0      1   
152                                   Meo, Mr. Alfonzo    male  55.5      0   
153                    van Billiard, Mr. Austin Blyler    male  40.5      0   
154                              Olsen, Mr. Ole Martin    male   NaN      0   
155                        Williams, Mr. Charles Duane    male  51.0      0   

     Parch            Ticket     Fare Cabin Embarked  newcol  
0        0         A/5 21171   7.2500   NaN        S  class3  
1        0          PC 17599  71.2833   C85        C  class1  
2        0  STON/O2. 3101282   7.9250   NaN        S  class3  
3        0            113803  53.1000  C123        S  class1  
4        0            373450   8.0500   NaN        S  class3  
..     ...               ...      ...   ...      ...     ...  
151      0            113776  66.6000    C2        S  class1  
152      0        A.5. 11206   8.0500   NaN        S  class3  
153      2          A/5. 851  14.5000   NaN        S  class3  
154      0         Fa 265302   7.3125   NaN        S  class3  
155      1          PC 17597  61.3792   NaN        C  class1  

[156 rows x 13 columns]

In [127]:
index=np.where(titanic_train["Fare"] == max(titanic_train["Fare"])) #maximum fare 
titanic_train.loc[index]               

PassengerId  Survived  Pclass                            Name     Sex  \
27           28         0       1  Fortune, Mr. Charles Alexander    male   
88           89         1       1      Fortune, Miss. Mabel Helen  female   

     Age  SibSp  Parch Ticket   Fare        Cabin Embarked  newcol  
27  19.0      3      2  19950  263.0  C23 C25 C27        S  class1  
88  23.0      3      2  19950  263.0  C23 C25 C27        S  class1

In [128]:
max(titanic_train["Fare"])

263.0

In [129]:
titanic_train["Fare"]==max(titanic_train["Fare"])

0      False
1      False
2      False
3      False
4      False
       ...  
151    False
152    False
153    False
154    False
155    False
Name: Fare, Length: 156, dtype: bool

In [130]:
index=np.where(titanic_train["Fare"] == max(titanic_train["Fare"])) #who has paid maximum fare 
titanic_train.loc[index]["Name"]            

27    Fortune, Mr. Charles Alexander
88        Fortune, Miss. Mabel Helen
Name: Name, dtype: object

In [ ]:
#here i am trying to merge a coulumn

In [132]:
titanic_train["Family"] = titanic_train["SibSp"] + titanic_train["Parch"]
titanic_train["Family"]
most_family = np.where(titanic_train["Family"] == max(titanic_train["Family"]))
most_family

(array([59, 71]),)

In [133]:
titanic_train["Family"] = titanic_train["SibSp"] + titanic_train["Parch"]

most_family = np.where(titanic_train["Family"] == max(titanic_train["Family"]))
most_family
titanic_train.iloc[most_family]

PassengerId  Survived  Pclass                                Name     Sex  \
59           60         0       3  Goodwin, Master. William Frederick    male   
71           72         0       3          Goodwin, Miss. Lillian Amy  female   

     Age  SibSp  Parch   Ticket  Fare Cabin Embarked  newcol  Family  
59  11.0      5      2  CA 2144  46.9   NaN        S  class3       7  
71  16.0      5      2  CA 2144  46.9   NaN        S  class3       7

In [139]:
# Data conversion operation
import numpy as np
import pandas as pd

labels = ['a','b','c','d']
my_data = [30,40,35,67]
arr = np.array(my_data)  
d = {'a':10,'b':20,'c':30}

print ("Labels:", labels)
print("My data:", my_data)
print("Dictionary:", d)
pd.Series(my_data, index=labels)

Labels: ['a', 'b', 'c', 'd']
My data: [30, 40, 35, 67]
Dictionary: {'a': 10, 'b': 20, 'c': 30}


a    30
b    40
c    35
d    67
dtype: int64

In [140]:
pd.Series(my_data, index=labels)

a    30
b    40
c    35
d    67
dtype: int64

In [145]:
print ("\nHolding numerical data\n",'-'*25, sep='')
print(pd.Series(labels))



Holding numerical data
-------------------------
0    a
1    b
2    c
3    d
dtype: object


In [141]:
print ("\nHolding numerical data\n",'-'*25, sep='')
print(pd.Series(arr)[1])



Holding numerical data
-------------------------
40


In [146]:
d

{'a': 10, 'b': 20, 'c': 30}

In [148]:
d.keys()

dict_keys(['a', 'b', 'c'])

In [150]:
d = {'a':"Puru",'b':20,'c':30}
d.items
pd.DataFrame(d,index = ['a','b','c'])

a   b   c
a  Puru  20  30
b  Puru  20  30
c  Puru  20  30

In [155]:
ser1 = pd.Series([1,2,3,4],index = [2,4,6,8])
ser2 = pd.Series([1,2,5,4],['CA', 'OR', 'NV', 'AZ'])
ser1

2    1
4    2
6    3
8    4
dtype: int64

In [158]:
ser1[0:3:2]

2    1
6    3
dtype: int64

In [159]:

from numpy.random import randn as rn
#np.random.seed(101)
matrix_data = rn(5,4)
row_labels = ['A','B','C','D','E']
column_headings = ['W','X','Y','Z']

df = pd.DataFrame(matrix_data,row_labels,column_headings)
#print("\nThe data frame looks like\n",'-'*45, sep='')
df

W         X         Y         Z
A  1.360294 -1.782414 -0.368801 -0.165383
B  0.050050  1.851773 -0.015774  0.037937
C -2.038415 -1.240874  0.253959 -0.874561
D  0.469838 -0.758534 -4.181314 -0.075498
E -0.402875  1.664289 -2.484149  0.310888

In [160]:
df.iloc[2]

W   -2.038415
X   -1.240874
Y    0.253959
Z   -0.874561
Name: C, dtype: float64

In [161]:
print("\nType of the pair of columns: ", type(df[['X','Z']]), sep='')
print ("\nSo, for more than one column, the object turns into a DataFrame")


Type of the pair of columns: <class 'pandas.core.frame.DataFrame'>

So, for more than one column, the object turns into a DataFrame


In [162]:
print("\nThe 'X' column accessed by DOT method (NOT recommended)\n",'-'*55, sep='')
print(df["X"])


The 'X' column accessed by DOT method (NOT recommended)
-------------------------------------------------------
A   -1.782414
B    1.851773
C   -1.240874
D   -0.758534
E    1.664289
Name: X, dtype: float64


In [163]:
print("\nA column is created by assigning it in relation to an existing column\n",'-'*75, sep='')
df['New'] = df['X']+df['Z']
df['New (Sum of X and Z)'] = df['X']+df['Z']
print(df)



A column is created by assigning it in relation to an existing column
---------------------------------------------------------------------------
          W         X         Y         Z       New  New (Sum of X and Z)
A  1.360294 -1.782414 -0.368801 -0.165383 -1.947797             -1.947797
B  0.050050  1.851773 -0.015774  0.037937  1.889710              1.889710
C -2.038415 -1.240874  0.253959 -0.874561 -2.115435             -2.115435
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032             -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177              1.975177


In [164]:
df

W         X         Y         Z       New  New (Sum of X and Z)
A  1.360294 -1.782414 -0.368801 -0.165383 -1.947797             -1.947797
B  0.050050  1.851773 -0.015774  0.037937  1.889710              1.889710
C -2.038415 -1.240874  0.253959 -0.874561 -2.115435             -2.115435
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032             -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177              1.975177

In [165]:
print("\nA column is dropped by using df.drop() method\n",'-'*55, sep=' ')
df.drop("C",inplace=True) # Notice the axis=1 option, axis = 0 is default, so one has to change it to 1




A column is dropped by using df.drop() method
 -------------------------------------------------------


In [166]:
df

W         X         Y         Z       New  New (Sum of X and Z)
A  1.360294 -1.782414 -0.368801 -0.165383 -1.947797             -1.947797
B  0.050050  1.851773 -0.015774  0.037937  1.889710              1.889710
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032             -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177              1.975177

In [167]:
df1=df.drop('A')
print("\nA row (index) is dropped by using df.drop() method and axis=0\n",'-'*65, sep='')
print(df1)



A row (index) is dropped by using df.drop() method and axis=0
-----------------------------------------------------------------
          W         X         Y         Z       New  New (Sum of X and Z)
B  0.050050  1.851773 -0.015774  0.037937  1.889710              1.889710
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032             -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177              1.975177


In [168]:
print("\nAn in-place change can be done by making inplace=True in the drop method\n",'-'*75, sep='')
df.drop('New (Sum of X and Z)', axis=1, inplace=True)
print(df)


An in-place change can be done by making inplace=True in the drop method
---------------------------------------------------------------------------
          W         X         Y         Z       New
A  1.360294 -1.782414 -0.368801 -0.165383 -1.947797
B  0.050050  1.851773 -0.015774  0.037937  1.889710
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177


In [169]:
### Selecting/indexing Rows
#* Label-based 'loc' method
#* Index (numeric) 'iloc' method
df

W         X         Y         Z       New
A  1.360294 -1.782414 -0.368801 -0.165383 -1.947797
B  0.050050  1.851773 -0.015774  0.037937  1.889710
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032
E -0.402875  1.664289 -2.484149  0.310888  1.975177

In [170]:
print("\nLabel-based 'loc' method can be used for selecting row(s)\n",'-'*60, sep='')
print("\nSingle row\n")
print(df.iloc[3])



Label-based 'loc' method can be used for selecting row(s)
------------------------------------------------------------

Single row

W     -0.402875
X      1.664289
Y     -2.484149
Z      0.310888
New    1.975177
Name: E, dtype: float64


In [172]:
print("\nMultiple rows\n")
print(df.loc[['B','C']])


Multiple rows



KeyError: 'Passing list-likes to .loc or [] with any missing labels is no longer supported, see https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#deprecate-loc-reindex-listlike'

In [173]:
print("\nIndex position based 'iloc' method can be used for selecting row(s)\n",'-'*70, sep='')
print("\nSingle row\n")
print(df.iloc[2])


Index position based 'iloc' method can be used for selecting row(s)
----------------------------------------------------------------------

Single row

W      0.469838
X     -0.758534
Y     -4.181314
Z     -0.075498
New   -0.834032
Name: D, dtype: float64


In [174]:
print("\nMultiple rows\n")
print(df.iloc[[1,2]])


Multiple rows

          W         X         Y         Z       New
B  0.050050  1.851773 -0.015774  0.037937  1.889710
D  0.469838 -0.758534 -4.181314 -0.075498 -0.834032


In [175]:
#### Subsetting DataFrame

In [176]:
matrix_data = rn(5,4)
row_labels = ['A','B','C','D','E']
column_headings = ['W','X','Y','Z']
df = pd.DataFrame(data=matrix_data, index=row_labels, columns=column_headings)



In [177]:
print("\nThe DatFrame\n",'-'*45, sep='')
print(df)



The DatFrame
---------------------------------------------
          W         X         Y         Z
A -0.769911 -0.739782 -0.918680 -0.578155
B  1.695844  0.704681  0.569274 -0.375063
C  1.904716  0.206187 -0.317210 -0.905898
D -0.984576  0.785634  1.082754 -0.871609
E  0.347803 -0.038427 -1.936502  0.732207


In [178]:
print("\nElement at row 'B' and column 'Y' is\n") 
print(df.loc[['B','C'],['Y','W']])



Element at row 'B' and column 'Y' is

          Y         W
B  0.569274  1.695844
C -0.317210  1.904716


In [179]:
print("\nSubset comprising of rows B and D, and columns W and Y, is\n")
df.iloc[[1,2,3],[0,1]]


Subset comprising of rows B and D, and columns W and Y, is



W         X
B  1.695844  0.704681
C  1.904716  0.206187
D -0.984576  0.785634

In [180]:
print(df.loc[['B','C'],])

          W         X         Y         Z
B  1.695844  0.704681  0.569274 -0.375063
C  1.904716  0.206187 -0.317210 -0.905898


In [181]:
print("\nThe DataFrame\n",'-'*45, sep='')
print(df)


The DataFrame
---------------------------------------------
          W         X         Y         Z
A -0.769911 -0.739782 -0.918680 -0.578155
B  1.695844  0.704681  0.569274 -0.375063
C  1.904716  0.206187 -0.317210 -0.905898
D -0.984576  0.785634  1.082754 -0.871609
E  0.347803 -0.038427 -1.936502  0.732207


In [182]:
print("\nBoolean DataFrame(s) where we are checking if the values are greater than 0\n",'-'*75, sep='')
print(df>0)



Boolean DataFrame(s) where we are checking if the values are greater than 0
---------------------------------------------------------------------------
       W      X      Y      Z
A  False  False  False  False
B   True   True   True  False
C   True   True  False  False
D  False   True   True  False
E   True  False  False   True


In [183]:
print("\n")
print(df.loc[['A','B','C']]>0)




       W      X      Y      Z
A  False  False  False  False
B   True   True   True  False
C   True   True  False  False


In [184]:
booldf = df>0
print("\nDataFrame indexed by boolean dataframe\n",'-'*45, sep='')
print(df[booldf])


DataFrame indexed by boolean dataframe
---------------------------------------------
          W         X         Y         Z
A       NaN       NaN       NaN       NaN
B  1.695844  0.704681  0.569274       NaN
C  1.904716  0.206187       NaN       NaN
D       NaN  0.785634  1.082754       NaN
E  0.347803       NaN       NaN  0.732207


In [185]:
import pandas as pd 
import numpy as np
matrix_data = np.matrix('22,66,140;42,70,148;30,62,125;35,68,160;25,62,152')
row_labels = ['A','B','C','D','E']
column_headings = ['Age', 'Height', 'Weight']
matrix_data

matrix([[ 22,  66, 140],
        [ 42,  70, 148],
        [ 30,  62, 125],
        [ 35,  68, 160],
        [ 25,  62, 152]])

In [186]:
df = pd.DataFrame(data=matrix_data, index=row_labels, columns=column_headings)
print("\nA new DataFrame\n",'-'*25, sep='')
print(df)



A new DataFrame
-------------------------
   Age  Height  Weight
A   22      66     140
B   42      70     148
C   30      62     125
D   35      68     160
E   25      62     152


In [187]:
df[df['Height']>65]

Age  Height  Weight
A   22      66     140
B   42      70     148
D   35      68     160

In [188]:
print("\nRows with Height > 65 inch\n",'-'*35, sep='')
print(df[df['Height']>65])



Rows with Height > 65 inch
-----------------------------------
   Age  Height  Weight
A   22      66     140
B   42      70     148
D   35      68     160


In [189]:
df['Height']>65

A     True
B     True
C    False
D     True
E    False
Name: Height, dtype: bool

In [190]:
booldf1 = df['Height']>65
booldf2 = df['Weight']>145

In [191]:
print("\nRows with Height > 65 inch and Weight >145 lbs\n",'-'*55, sep='')
print(df[(booldf1) & (booldf2)])



Rows with Height > 65 inch and Weight >145 lbs
-------------------------------------------------------
   Age  Height  Weight
B   42      70     148
D   35      68     160


In [192]:
print("\nDataFrame with only Age and Weight columns whose Height > 65 inch\n",'-'*68, sep='')
print(df[booldf1][['Age','Weight']])


DataFrame with only Age and Weight columns whose Height > 65 inch
--------------------------------------------------------------------
   Age  Weight
A   22     140
B   42     148
D   35     160


# Data manipulation 2

In [1]:
import numpy as np
import pandas as pd
import os
import matplotlib.pyplot as plt
%matplotlib inline

In [4]:
({'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]})

{'A': [1, 2, nan], 'B': [5, nan, nan], 'C': [1, 2, 3]}

In [6]:
({'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]})
df = pd.DataFrame({'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]})
df

A    B  C
0  1.0  5.0  1
1  2.0  NaN  2
2  NaN  NaN  3

In [9]:
# add new column 
df["states"]="CA NV AZ" .split()
df

A    B  C states
0  1.0  5.0  1     CA
1  2.0  NaN  2     NV
2  NaN  NaN  3     AZ

In [10]:
#set index
df = pd.DataFrame({'A':[1,2,np.nan],'B':[5,np.nan,np.nan],'C':[1,2,3]})
df['States']="CA NV AZ".split()
df.set_index('States',inplace=True)
print(df)

          A    B  C
States             
CA      1.0  5.0  1
NV      2.0  NaN  2
AZ      NaN  NaN  3


In [11]:
# permanent set indexes
df["states"]="CA NV AZ" .split()
df.set_index('states', inplace = True)
df

A    B  C
states             
CA      1.0  5.0  1
NV      2.0  NaN  2
AZ      NaN  NaN  3

In [14]:
# if i have to drop a raw and coloumn so which is having a nan  value so what should i do 

In [13]:
print("\nDropping any rows with a NaN value\n",'-'*35, sep='')
print(df.dropna(axis=0))



Dropping any rows with a NaN value
-----------------------------------
          A    B  C
states             
CA      1.0  5.0  1


In [15]:
print("\nDropping any column with a NaN value\n",'-'*35, sep='')
print(df.dropna(axis=1))


Dropping any column with a NaN value
-----------------------------------
        C
states   
CA      1
NV      2
AZ      3


In [16]:
print("\nDropping a row with a minimum 2 NaN value using 'thresh' parameter\n",'-'*68, sep='')
print(df.dropna(axis=0, thresh=4))


Dropping a row with a minimum 2 NaN value using 'thresh' parameter
--------------------------------------------------------------------
Empty DataFrame
Columns: [A, B, C]
Index: []


In [17]:
df.dropna(axis=0 , thresh=2)

A    B  C
states             
CA      1.0  5.0  1
NV      2.0  NaN  2

In [22]:
df.dropna(axis=0 , thresh=3)

A    B  C
states             
CA      1.0  5.0  1

In [27]:
print("\nFilling values with a default value\n",'-'*35, sep='')
print(df.fillna(value='FILL VALUE'))



Filling values with a default value
-----------------------------------
                 A           B  C
states                           
CA               1           5  1
NV               2  FILL VALUE  2
AZ      FILL VALUE  FILL VALUE  3


In [28]:
print("\nFilling values with a computed value (mean of column A here)\n",'-'*60, sep='')
print(df.fillna(value=df['A'].mean()))


Filling values with a computed value (mean of column A here)
------------------------------------------------------------
          A    B  C
states             
CA      1.0  5.0  1
NV      2.0  1.5  2
AZ      1.5  1.5  3


In [32]:
# Create dataframe
data = {'Company':['GOOG','GOOG','MSFT','MSFT','FB','FB'],
       'Person':['Sam','Charlie','Amy','Vanessa','Carl','Sarah'],
       'Sales':[200,120,340,124,243,350]}
df = pd.DataFrame(data)
df

Company   Person  Sales
0    GOOG      Sam    200
1    GOOG  Charlie    120
2    MSFT      Amy    340
3    MSFT  Vanessa    124
4      FB     Carl    243
5      FB    Sarah    350

In [34]:
byComp =df.groupby('Company')
byComp
byComp.mean()

Sales
Company       
FB       296.5
GOOG     160.0
MSFT     232.0

In [30]:
# grouping operationthis is always returns an object
byComp = df.groupby('Company')
print("\nGrouping by 'Company' column and listing mean sales\n",'-'*55, sep='')
print(byComp.mean())



Grouping by 'Company' column and listing mean sales
-------------------------------------------------------
         Sales
Company       
FB       296.5
GOOG     160.0
MSFT     232.0


In [35]:
print("\nGrouping by 'Company' column and listing sum of sales\n",'-'*55, sep='')
print(byComp.sum())



Grouping by 'Company' column and listing sum of sales
-------------------------------------------------------
         Sales
Company       
FB         593
GOOG       320
MSFT       464


In [37]:
df.groupby('Company')

In [38]:
df.groupby('Company').describe()

Sales                                                        
        count   mean         std    min     25%    50%     75%    max
Company                                                              
FB        2.0  296.5   75.660426  243.0  269.75  296.5  323.25  350.0
GOOG      2.0  160.0   56.568542  120.0  140.00  160.0  180.00  200.0
MSFT      2.0  232.0  152.735065  124.0  178.00  232.0  286.00  340.0

In [39]:
n = df.groupby('Company').describe()
n

Sales                                                        
        count   mean         std    min     25%    50%     75%    max
Company                                                              
FB        2.0  296.5   75.660426  243.0  269.75  296.5  323.25  350.0
GOOG      2.0  160.0   56.568542  120.0  140.00  160.0  180.00  200.0
MSFT      2.0  232.0  152.735065  124.0  178.00  232.0  286.00  340.0

In [40]:
n['Sales']['count']

Company
FB      2.0
GOOG    2.0
MSFT    2.0
Name: count, dtype: float64

In [42]:
# how to extarct the data from raw
n = df.groupby('Company').describe().loc['FB'] 
n 


Sales  count      2.000000
       mean     296.500000
       std       75.660426
       min      243.000000
       25%      269.750000
       50%      296.500000
       75%      323.250000
       max      350.000000
Name: FB, dtype: float64

In [43]:
print("\nAll in one line of command (Stats for 'FB')\n",'-'*65, sep='')
print(pd.DataFrame(df.groupby('Company').describe().loc['FB']).transpose())



All in one line of command (Stats for 'FB')
-----------------------------------------------------------------
   Sales                                                       
   count   mean        std    min     25%    50%     75%    max
FB   2.0  296.5  75.660426  243.0  269.75  296.5  323.25  350.0


In [48]:
m=pd.DataFrame(df.groupby('Company').describe().loc['FB'] )
m

FB
Sales count    2.000000
      mean   296.500000
      std     75.660426
      min    243.000000
      25%    269.750000
      50%    296.500000
      75%    323.250000
      max    350.000000

In [50]:
m.loc['Sales'].loc["min"]

FB    243.0
Name: min, dtype: float64

In [51]:
m.loc['Sales'].loc["max"]

FB    350.0
Name: max, dtype: float64

In [52]:
m=pd.DataFrame(df.groupby('Company').describe().loc['FB'] ).transpose()
# transpose opeartion it will just try to rotate by 180 degree

In [54]:
m

Sales                                                       
   count   mean        std    min     25%    50%     75%    max
FB   2.0  296.5  75.660426  243.0  269.75  296.5  323.25  350.0

In [55]:
# Merging two data frames
# Creating data frames
df1 = pd.DataFrame({'A': ['A0', 'A1', 'A2', 'A3'],
                        'B': ['B0', 'B1', 'B2', 'B3'],
                        'C': ['C0', 'C1', 'C2', 'C3'],
                        'D': ['D0', 'D1', 'D2', 'D3']},
                        index=[0, 1, 2, 3])

In [56]:
df1

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [57]:
df2 = pd.DataFrame({'A': ['A4', 'A5', 'A6', 'A7'],
                        'B': ['B4', 'B5', 'B6', 'B7'],
                        'C': ['C4', 'C5', 'C6', 'C7'],
                        'D': ['D4', 'D5', 'D6', 'D7']},
                         index=[0, 1, 2, 3])


In [58]:
df2

A   B   C   D
0  A4  B4  C4  D4
1  A5  B5  C5  D5
2  A6  B6  C6  D6
3  A7  B7  C7  D7

In [59]:
df3 = pd.DataFrame({'A': ['A8', 'A9', 'A10', 'A11'],
                        'B': ['B8', 'B9', 'B10', 'B11'],
                        'C': ['C8', 'C9', 'C10', 'C11'],
                        'D': ['D8', 'D9', 'D10', 'D11']},
                        index=[8,9,10,11])


In [60]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [61]:
print("\nThe DataFrame number 1\n",'-'*30, sep='')
print(df1)



The DataFrame number 1
------------------------------
    A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3


In [62]:
print("\nThe DataFrame number 2\n",'-'*30, sep='')
print(df2)


The DataFrame number 2
------------------------------
    A   B   C   D
0  A4  B4  C4  D4
1  A5  B5  C5  D5
2  A6  B6  C6  D6
3  A7  B7  C7  D7


In [63]:
print("\nThe DataFrame number 3\n",'-'*30, sep='')
print(df3)


The DataFrame number 3
------------------------------
      A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11


In [73]:
#concatenation
df_cat1 = pd.concat([df1,df2,df3], axis=0)
print("\nAfter concatenation along row\n",'-'*30, sep='')
print(df_cat1)
df_cat1.loc[2]


After concatenation along row
------------------------------
      A    B    C    D
0    A0   B0   C0   D0
1    A1   B1   C1   D1
2    A2   B2   C2   D2
3    A3   B3   C3   D3
0    A4   B4   C4   D4
1    A5   B5   C5   D5
2    A6   B6   C6   D6
3    A7   B7   C7   D7
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11


A   B   C   D
2  A2  B2  C2  D2
2  A6  B6  C6  D6

In [71]:
df_cat1.iloc[4]

A    A4
B    B4
C    C4
D    D4
Name: 0, dtype: object

In [74]:
df_cat2 = pd.concat([df1,df2,df3], axis=1)
print("\nAfter concatenation along column\n",'-'*60, sep='')
print(df_cat2)



After concatenation along column
------------------------------------------------------------
      A    B    C    D    A    B    C    D    A    B    C    D
0    A0   B0   C0   D0   A4   B4   C4   D4  NaN  NaN  NaN  NaN
1    A1   B1   C1   D1   A5   B5   C5   D5  NaN  NaN  NaN  NaN
2    A2   B2   C2   D2   A6   B6   C6   D6  NaN  NaN  NaN  NaN
3    A3   B3   C3   D3   A7   B7   C7   D7  NaN  NaN  NaN  NaN
8   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A8   B8   C8   D8
9   NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN   A9   B9   C9   D9
10  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A10  B10  C10  D10
11  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  A11  B11  C11  D11


In [75]:
df_cat2.fillna(value=0, inplace=True)
print("\nAfter filling missing values with zero\n",'-'*60, sep='')
print(df_cat2)


After filling missing values with zero
------------------------------------------------------------
     A   B   C   D   A   B   C   D    A    B    C    D
0   A0  B0  C0  D0  A4  B4  C4  D4    0    0    0    0
1   A1  B1  C1  D1  A5  B5  C5  D5    0    0    0    0
2   A2  B2  C2  D2  A6  B6  C6  D6    0    0    0    0
3   A3  B3  C3  D3  A7  B7  C7  D7    0    0    0    0
8    0   0   0   0   0   0   0   0   A8   B8   C8   D8
9    0   0   0   0   0   0   0   0   A9   B9   C9   D9
10   0   0   0   0   0   0   0   0  A10  B10  C10  D10
11   0   0   0   0   0   0   0   0  A11  B11  C11  D11


In [76]:
# merging by a common key

In [77]:
left = pd.DataFrame({'key': ['K0', 'K8', 'K2', 'K3'],
                     'A': ['A0', 'A1', 'A2', 'A3'],
                     'B': ['B0', 'B1', 'B2', 'B3']})
   
right = pd.DataFrame({'key': ['K0', 'K1', 'K2', 'K3'],
                          'C': ['C0', 'C1', 'C2', 'C3'],
                          'D': ['D0', 'D1', 'D2', 'D3']})


In [78]:
left

key   A   B
0  K0  A0  B0
1  K8  A1  B1
2  K2  A2  B2
3  K3  A3  B3

In [79]:
right

key   C   D
0  K0  C0  D0
1  K1  C1  D1
2  K2  C2  D2
3  K3  C3  D3

In [80]:
print("\nThe DataFrame 'left'\n",'-'*30, sep='')
print(left)



The DataFrame 'left'
------------------------------
  key   A   B
0  K0  A0  B0
1  K8  A1  B1
2  K2  A2  B2
3  K3  A3  B3


In [81]:
print("\nThe DataFrame 'right'\n",'-'*30, sep='')
print(right)


The DataFrame 'right'
------------------------------
  key   C   D
0  K0  C0  D0
1  K1  C1  D1
2  K2  C2  D2
3  K3  C3  D3


In [82]:
merge1= pd.merge(left,right,how='inner',on='key')
print("\nAfter simple merging with 'inner' method\n",'-'*50, sep='')
print(merge1)


After simple merging with 'inner' method
--------------------------------------------------
  key   A   B   C   D
0  K0  A0  B0  C0  D0
1  K2  A2  B2  C2  D2
2  K3  A3  B3  C3  D3


In [83]:
left = pd.DataFrame({'key1': ['K0', 'K0', 'K1', 'K2'],
                     'key2': ['K0', 'K1', 'K0', 'K1'],
                        'A': ['A0', 'A1', 'A2', 'A3'],
                        'B': ['B0', 'B1', 'B2', 'B3']})
    
right = pd.DataFrame({'key1': ['K0', 'K1', 'K1', 'K2'],
                               'key2': ['K0', 'K0', 'K0', 'K0'],
                                  'C': ['C0', 'C1', 'C2', 'C3'],
                                  'D': ['D0', 'D1', 'D2', 'D3']})

In [84]:
left

key1 key2   A   B
0   K0   K0  A0  B0
1   K0   K1  A1  B1
2   K1   K0  A2  B2
3   K2   K1  A3  B3

In [85]:
right

key1 key2   C   D
0   K0   K0  C0  D0
1   K1   K0  C1  D1
2   K1   K0  C2  D2
3   K2   K0  C3  D3

In [86]:
pd.merge(left, right, on=['key1', 'key2'])

key1 key2   A   B   C   D
0   K0   K0  A0  B0  C0  D0
1   K1   K0  A2  B2  C1  D1
2   K1   K0  A2  B2  C2  D2

In [87]:
pd.merge(left, right, how='left',on=['key1', 'key2'])

key1 key2   A   B    C    D
0   K0   K0  A0  B0   C0   D0
1   K0   K1  A1  B1  NaN  NaN
2   K1   K0  A2  B2   C1   D1
3   K1   K0  A2  B2   C2   D2
4   K2   K1  A3  B3  NaN  NaN

In [88]:
pd.merge(left, right, how='right',on=['key1', 'key2'])

key1 key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K1   K0   A2   B2  C1  D1
2   K1   K0   A2   B2  C2  D2
3   K2   K0  NaN  NaN  C3  D3

In [89]:
pd.merge(left, right, how='right',on=['key1', 'key2'])

key1 key2    A    B   C   D
0   K0   K0   A0   B0  C0  D0
1   K1   K0   A2   B2  C1  D1
2   K1   K0   A2   B2  C2  D2
3   K2   K0  NaN  NaN  C3  D3

In [90]:
left

key1 key2   A   B
0   K0   K0  A0  B0
1   K0   K1  A1  B1
2   K1   K0  A2  B2
3   K2   K1  A3  B3

In [91]:
right

key1 key2   C   D
0   K0   K0  C0  D0
1   K1   K0  C1  D1
2   K1   K0  C2  D2
3   K2   K0  C3  D3

In [98]:
# Define a function
def testfunc(x):
    if (x> 500):
        return (10*np.log10(x))
    else:
        return (x/10)

In [99]:
df = pd.DataFrame({'col1':[1,2,3,4,5,6,7,8,9,10],
                   'col2':[444,555,666,444,333,222,666,777,666,555],
                   'col3':'aaa bb c dd eeee fff gg h iii j'.split()})
df

col1  col2  col3
0     1   444   aaa
1     2   555    bb
2     3   666     c
3     4   444    dd
4     5   333  eeee
5     6   222   fff
6     7   666    gg
7     8   777     h
8     9   666   iii
9    10   555     j

In [104]:
df['col2'].apply(testfunc)

0    44.400000
1    27.442930
2    28.234742
3    44.400000
4    33.300000
5    22.200000
6    28.234742
7    28.904210
8    28.234742
9    27.442930
Name: col2, dtype: float64

In [106]:
df['new']=df['col2'].apply(testfunc)
df

col1  col2  col3  FuncApplied  col3length        new
0     1   444   aaa     6.095825           3  44.400000
1     2   555    bb     6.318968           2  27.442930
2     3   666     c     6.501290           1  28.234742
3     4   444    dd     6.095825           2  44.400000
4     5   333  eeee     5.808142           4  33.300000
5     6   222   fff     5.402677           3  22.200000
6     7   666    gg     6.501290           2  28.234742
7     8   777     h     6.655440           1  28.904210
8     9   666   iii     6.501290           3  28.234742
9    10   555     j     6.318968           1  27.442930

In [100]:
df['FuncApplied'] = df['col2'].apply(lambda x : np.log(x))
print(df)

   col1  col2  col3  FuncApplied
0     1   444   aaa     6.095825
1     2   555    bb     6.318968
2     3   666     c     6.501290
3     4   444    dd     6.095825
4     5   333  eeee     5.808142
5     6   222   fff     5.402677
6     7   666    gg     6.501290
7     8   777     h     6.655440
8     9   666   iii     6.501290
9    10   555     j     6.318968


In [101]:
df['col3length']= df['col3'].apply(len)
print(df)

   col1  col2  col3  FuncApplied  col3length
0     1   444   aaa     6.095825           3
1     2   555    bb     6.318968           2
2     3   666     c     6.501290           1
3     4   444    dd     6.095825           2
4     5   333  eeee     5.808142           4
5     6   222   fff     5.402677           3
6     7   666    gg     6.501290           2
7     8   777     h     6.655440           1
8     9   666   iii     6.501290           3
9    10   555     j     6.318968           1


In [102]:
df['FuncApplied'].apply(lambda x: np.sqrt(x))

0    2.468972
1    2.513756
2    2.549763
3    2.468972
4    2.410009
5    2.324366
6    2.549763
7    2.579814
8    2.549763
9    2.513756
Name: FuncApplied, dtype: float64

In [103]:
print("\nSum of the column 'FuncApplied' is: ",df['FuncApplied'].sum())


Sum of the column 'FuncApplied' is:  62.19971458619886


In [107]:
print("Mean of the column 'FuncApplied' is: ",df['FuncApplied'].mean())


Mean of the column 'FuncApplied' is:  6.219971458619886


In [108]:
print("Std dev of the column 'FuncApplied' is: ",df['FuncApplied'].std())


Std dev of the column 'FuncApplied' is:  0.3822522801574853


In [109]:
print("Min and max of the column 'FuncApplied' are: ",df['FuncApplied'].min(),"and",df['FuncApplied'].max())

Min and max of the column 'FuncApplied' are:  5.402677381872279 and 6.655440350367647


In [110]:
### Deletion, sorting, list of column and row names

In [111]:
print("\nName of columns\n",'-'*20, sep='')
print(df.columns)



Name of columns
--------------------
Index(['col1', 'col2', 'col3', 'FuncApplied', 'col3length', 'new'], dtype='object')


In [112]:
l = list(df.columns)
print("\nColumn names in a list of strings for later manipulation:",l)


Column names in a list of strings for later manipulation: ['col1', 'col2', 'col3', 'FuncApplied', 'col3length', 'new']


In [113]:
print("\nDeleting last column by 'del' command\n",'-'*50, sep='')
del df['col3length']
print(df)
df['col3length']= df['col3'].apply(len)


Deleting last column by 'del' command
--------------------------------------------------
   col1  col2  col3  FuncApplied        new
0     1   444   aaa     6.095825  44.400000
1     2   555    bb     6.318968  27.442930
2     3   666     c     6.501290  28.234742
3     4   444    dd     6.095825  44.400000
4     5   333  eeee     5.808142  33.300000
5     6   222   fff     5.402677  22.200000
6     7   666    gg     6.501290  28.234742
7     8   777     h     6.655440  28.904210
8     9   666   iii     6.501290  28.234742
9    10   555     j     6.318968  27.442930


In [114]:
df.sort_values(by='col2') #inplace=False by default

col1  col2  col3  FuncApplied        new  col3length
5     6   222   fff     5.402677  22.200000           3
4     5   333  eeee     5.808142  33.300000           4
0     1   444   aaa     6.095825  44.400000           3
3     4   444    dd     6.095825  44.400000           2
1     2   555    bb     6.318968  27.442930           2
9    10   555     j     6.318968  27.442930           1
2     3   666     c     6.501290  28.234742           1
6     7   666    gg     6.501290  28.234742           2
8     9   666   iii     6.501290  28.234742           3
7     8   777     h     6.655440  28.904210           1

In [115]:
df.sort_values(by='FuncApplied',ascending=False) #inplace=False by default

col1  col2  col3  FuncApplied        new  col3length
7     8   777     h     6.655440  28.904210           1
2     3   666     c     6.501290  28.234742           1
6     7   666    gg     6.501290  28.234742           2
8     9   666   iii     6.501290  28.234742           3
1     2   555    bb     6.318968  27.442930           2
9    10   555     j     6.318968  27.442930           1
0     1   444   aaa     6.095825  44.400000           3
3     4   444    dd     6.095825  44.400000           2
4     5   333  eeee     5.808142  33.300000           4
5     6   222   fff     5.402677  22.200000           3

In [116]:
df = pd.DataFrame({'col1':[1,2,3,np.nan],
                   'col2':[None,555,666,444],
                   'col3':['abc','def','ghi','xyz']})
df.head()

col1   col2 col3
0   1.0    NaN  abc
1   2.0  555.0  def
2   3.0  666.0  ghi
3   NaN  444.0  xyz

In [117]:
df.isnull()

col1   col2   col3
0  False   True  False
1  False  False  False
2  False  False  False
3   True  False  False

In [118]:
df.fillna('FILL')

col1  col2 col3
0     1  FILL  abc
1     2   555  def
2     3   666  ghi
3  FILL   444  xyz

In [119]:
df1


A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3

In [120]:
df2

A   B   C   D
0  A4  B4  C4  D4
1  A5  B5  C5  D5
2  A6  B6  C6  D6
3  A7  B7  C7  D7

In [121]:
df3

A    B    C    D
8    A8   B8   C8   D8
9    A9   B9   C9   D9
10  A10  B10  C10  D10
11  A11  B11  C11  D11

In [122]:
pd.merge(df1, df2, how='inner')

Empty DataFrame
Columns: [A, B, C, D]
Index: []

In [123]:
pd.merge(df1, df2, how='outer')

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3
4  A4  B4  C4  D4
5  A5  B5  C5  D5
6  A6  B6  C6  D6
7  A7  B7  C7  D7

In [124]:
pd.merge(df1, df2, how='left')

A   B   C   D
0  A0  B0  C0  D0
1  A1  B1  C1  D1
2  A2  B2  C2  D2
3  A3  B3  C3  D3